File: ios/localasset-core/Sources/LocalAssetCore/internal/ResolverChain.swift
public struct ResolverChain {
    private let resolvers: [ResourceResolver]

    public init(resolvers: [ResourceResolver]) {
        self.resolvers = resolvers
    }

    public func resolve(
        request: AssetRequest,
        context: ResolveContext,
        observer: EngineObserver? = nil
    ) -> ResolverResult {
        for (index, resolver) in resolvers.enumerated() {
            let result = resolver.resolve(request: request, context: context)
            let descriptor: ResourceDescriptor? = {
                if case .hit(let d) = result { return d }
                return nil
            }()
            observer?.onStage(EngineStageEvent(
                stage: EngineStage.resolve,
                request: request,
                resolverIndex: index,
                resolver: resolver,
                resolverResult: result,
                descriptor: descriptor
            ))
            switch result {
            case .hit, .failure: return result
            case .skip: continue
            }
        }
        return .skip
    }
}

File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/internal/ResolverChain.kt
package io.github.xesam.android.localasset.core.internal

import io.github.xesam.android.localasset.core.api.EngineObserver
import io.github.xesam.android.localasset.core.api.EngineStageEvent
import io.github.xesam.android.localasset.core.api.ResolverResult
import io.github.xesam.android.localasset.core.api.ResourceResolver
import io.github.xesam.android.localasset.core.model.AssetRequest
import io.github.xesam.android.localasset.core.model.ResolveContext

class ResolverChain(
    private val resolvers: List<ResourceResolver>,
) {
    fun resolve(
        request: AssetRequest,
        context: ResolveContext,
        observer: EngineObserver? = null,
    ): ResolverResult {
        resolvers.forEachIndexed { index, resolver ->
            val result = resolver.resolve(request, context)
            observer?.onStage(
                EngineStageEvent(
                    stage = EngineObserver.STAGE_RESOLVE,
                    request = request,
                    resolverIndex = index,
                    resolver = resolver,
                    resolverResult = result,
                    descriptor = (result as? ResolverResult.Hit)?.descriptor,
                ),
            )
            when (result) {
                is ResolverResult.Hit -> return result
                is ResolverResult.Failure -> return result
                ResolverResult.Skip -> Unit
            }
        }
        return ResolverResult.Skip
    }
}
